File: LINKED LIST/doublycircular.c
#include<stdio.h>
#include<stdlib.h>
struct node
{
    int data;
    struct node * next;
    struct node * prev;
};
typedef struct node NODE;
typedef struct node* PNODE;
typedef struct node** PPNODE;
void Display(PNODE first , PNODE last)
{
    printf("<=>");
    do
    {
        printf("|%d|<=>",first->data);
        first=first->next;
    } while (first!=last->next);
}
int Count(PNODE first , PNODE last)
{
    int iCount = 0;
    do
    {
        iCount++;
        first=first->next;
    } while (first!=last->next);
    return iCount;
}
void insertFirst(PPNODE first , PPNODE last , int iNo)
{
    PNODE newn = NULL;
    newn = (PNODE)malloc(sizeof(NODE));
    newn->data=iNo;
    newn->next=NULL;
    newn->prev=NULL;

    if(*first==NULL && *last == NULL)
    {
        *first=newn;
        *last = newn;
    }
    else
    {
        newn->next= *first;
        *first = newn;
        newn->next->prev=newn;
    }

    (*last)->next = *first;
    (*first)->prev=*last;
}

void insertLast(PPNODE first , PPNODE last , int iNo)
{
    PNODE newn = NULL;
    newn = (PNODE)malloc(sizeof(NODE));
    newn->data=iNo;
    newn->next=NULL;
    newn->prev=NULL;
    
    if(*first==NULL && *last == NULL)
    {
        *first=newn;
        *last = newn;   
    }
    else
    {
        (*last)->next = newn;
        newn->prev=*last;
        *last = newn;
    }
      (*last)->next = *first;
      (*first)->prev=*last;
}

void insertAtpos(PPNODE first , PPNODE last , int iNo , int pos)
{
    int icount = Count(*first,*last);
    if(pos<1||pos>icount+1)
    {
        return;
    }
    if(pos==1)
    {
      insertFirst(first,last,iNo);
    }
    else if(pos==icount+1)
    {
       insertLast(first,last,iNo);
    }
    else
    {
    PNODE newn = NULL;
    newn = (PNODE)malloc(sizeof(NODE));
    newn->data=iNo;
    newn->next=NULL;
    newn->prev=NULL;

     PNODE temp = NULL;
     temp=*first;

     for(int iCnt = 1 ; iCnt<=pos-2 ; iCnt++)
     {
        temp=temp->next;
     }

     newn->next=temp->next;
     newn->next->prev=newn;
     temp->next=newn;
     newn->prev=temp;
     }

}

void DeleteFirst(PPNODE first , PPNODE last)
{
    if(*first==NULL && *last==NULL)
    {
         return ;
    }
    else
    {
        *first=(*first)->next;
        free((*last)->next);
        (*last)->next=*first;
        (*first)->prev=*last;
    }

}
void DeleteLast(PPNODE first , PPNODE last )
{
    if(*first==NULL && *last==NULL)
    {
         return;
    }
    else
    {
        PNODE temp = NULL;
        temp = *first;

        while (temp->next!=*last)
        {
            temp=temp->next;
        }
        free(temp->next);
        *last=temp;
        (*last)->next=*first;
        (*first)->prev=*last;
    }
}

void DeleteAtpos(PPNODE first , PPNODE last, int pos)
{
    int icount = Count(*first,*last);
    if(pos<1||pos>icount+1)
    {
        return;
    }
    if(pos==1)
    {
      DeleteFirst(first , last);
    }
    else if(pos==icount)
    {
       DeleteLast(first,last);
    }
    else
    {
       PNODE temp = *first;
       PNODE target = (*first)->next;
       for(int iCnt = 1; iCnt<=pos-2 ; iCnt++)
       {
         temp = temp ->next;
         target = target ->next;
       }

       temp->next=target->next;
       target->next->prev=temp;
       free(target);
    }
}

int main()
{
    PNODE Head = NULL;
    PNODE Tail = NULL;
    int iRet = 0;
    int iNo = 0;
    int iPos = 0;
    int ch = 0;
    do{
    printf("\n");
    printf("------------MENU------------------\n");
    printf("1.INSERT AT FIRST POSITION\n");
    printf("2.INSERT AT LAST POSITON\n");
    printf("3.INSERT AT ANY POSITION\n");
    printf("4.DELETE AT FIRST POSITION\n");
    printf("5.DELETE AT LAST POSITON\n");
    printf("6.DELETE AT ANY POSITION\n");
    printf("7.DISPLAY \n");
    printf("8.COUNT\n");
    printf("9.EXIT");
    printf("\n");

    printf("PLEASE ENTER YOUR CHOICE \n");
    scanf("%d",&ch);

    if(ch<1||ch>9)
    {
        printf("INVALID INPUT");
    }
    switch (ch)
    {
       case 1:
       printf("ENTER NUMBER YOU WANT TO INSERT:\n");
       scanf("%d",&iNo);
       insertFirst(&Head,&Tail,iNo);
       break;
       case 2:
       printf("ENTER NUMBER YOU WANT TO INSERT:\n");
       scanf("%d",&iNo);
       insertLast(&Head,&Tail,iNo);
       break;
       case 3:
       printf("ENTER NUMBER YOU WANT TO INSERT:\n");
       scanf("%d",&iNo);
       printf("ENTER POSITION :\n");
       scanf("%d",&iPos);
       insertAtpos(&Head,&Tail,iNo,iPos);
       break;
       case 4:
       DeleteFirst(&Head,&Tail);
       break;
       case 5:
       DeleteLast(&Head,&Tail);
       break;
       case 6:
       printf("ENTER POSITION TO DELETE :\n");
       scanf("%d",&iPos);
       DeleteAtpos(&Head,&Tail,iPos);
       break;
       case 7:
       Display(Head,Tail);
       break;
       case 8:
       iRet=Count(Head,Tail);
       printf("NUMBER OF ELEMENTS IN LINKED LIST : %d",iRet);
       break;
       case 9:
       printf("EXISTING PROGRAM");

    }
    }while (ch!=9);
    return 0;
    


    return 0;
}

File: LINKED LIST/doublycircular.java
import java.util.Scanner;

class node {
    public int data;
    public node next;
    public node prev;
}

class DoublyC {
    public node first;
    public node last;
    public int iCount;

    public DoublyC() {
        first = null;
        last = null;
        iCount = 0;
    }

    public void display() {
        if (first == null && last == null) {
            System.out.println("Linked list is empty");
            return;
        }
        node temp = first;
        System.out.println("<=>");
        do {
            System.out.print("|" + temp.data + "|<=>");
            temp = temp.next;
        } while (temp != first);
    }

    public void insertFirst(int no) {
        node newn = new node();
        newn.data = no;
        newn.next = null;
        newn.prev = null;
        if (first == null && last == null) {
            first = newn;
            last = newn;
        } else {
            newn.next = first;
            newn.next.prev = newn;
            first = newn;
        }
        last.next = first;
        first.prev = last;
        iCount++;
    }

    public void insertLast(int no) {
        node newn = new node();
        newn.data = no;
        newn.next = null;
        newn.prev = null;
        if (first == null && last == null) {
            first = newn;
            last = newn;
        } else {
            last.next = newn;
            newn.prev = last;
            last = newn;
        }
        last.next = first;
        first.prev = last;
        iCount++;
    }

    public void insertAtpos(int no, int pos) {
        if (pos < 1 || pos > iCount + 1) {
            return;
        }
        if (pos == 1) {
            insertFirst(no);
        } else if (pos == iCount + 1) {
            insertLast(no);
        } else {
            node newn = new node();
            newn.data = no;
            newn.next = null;
            newn.prev = null;
            node temp = first;
            for (int iCnt = 1; iCnt <= pos - 2; iCnt++) {
                temp = temp.next;
            }
            newn.next = temp.next;
            newn.next.prev = newn;
            temp.next = newn;
            newn.prev = temp;
            iCount++;
        }
    }

    public void deleteFirst() {
        if (first == null && last == null) {
            System.out.println("Linked list is empty");
            return;
        } else if (first == last) {
            first = null;
            last = null;
        } else {
            first = first.next;
            last.next = first;
            first.prev = last;
        }
        iCount--;
    }

    public void deleteLast() {
        if (first == null && last == null) {
            System.out.println("Linked list is empty");
            return;
        } else if (first == last) {
            first = null;
            last = null;
        } else {
            node temp = first;
            while (temp.next != last) {
                temp = temp.next;
            }
            last = temp;
            last.next = first;
            first.prev = last;
        }
        iCount--;
    }

    public void deleteAtpos(int pos) {
        if (pos < 1 || pos > iCount) {
            System.out.println("invalid position");
            return;
        }
        if (pos == 1) {
            deleteFirst();
        } else if (pos == iCount) {
            deleteLast();
        } else {
            node temp = first;
            for (int iCnt = 1; iCnt <= pos - 2; iCnt++) {
                temp = temp.next;
            }
            temp.next = temp.next.next;
            temp.next.prev = temp;
            iCount--;
        }
    }

    public int count() {
        return iCount;
    }
}

class DoublyCLL {
    public static void main(String args[]) {
        DoublyC sobj = new DoublyC();
        int iRet = 0;
        int iNo = 0;
        int iPos = 0;
        int ch = 0;
        Scanner s = new Scanner(System.in);
        do {
            System.out.println("------------MENU------------------");
            System.out.println("1.INSERT AT FIRST POSITION");
            System.out.println("2.INSERT AT LAST POSITON");
            System.out.println("3.INSERT AT ANY POSITION");
            System.out.println("4.DELETE AT FIRST POSITION");
            System.out.println("5.DELETE AT LAST POSITON");
            System.out.println("6.DELETE AT ANY POSITION");
            System.out.println("7.DISPLAY ");
            System.out.println("8.COUNT");
            System.out.println("9.EXIT");
            System.out.println("PLEASE ENTER YOUR CHOICE ");
            ch = s.nextInt();
            if (ch < 1 || ch > 9) {
                System.out.println("invalid input");
            }
            System.out.println();
            switch (ch) {

                case 1:
                    System.out.println("Enter Number you want to insert:");
                    iNo = s.nextInt();
                    sobj.insertFirst(iNo);
                    break;
                case 2:
                    System.out.println("Enter Number you want to insert:");
                    iNo = s.nextInt();
                    sobj.insertLast(iNo);
                    break;
                case 3:
                    System.out.println("Enter Number you want to insert:");
                    iNo = s.nextInt();
                    System.out.println("Enter position");
                    iPos = s.nextInt();
                    sobj.insertAtpos(iNo, iPos);
                    break;
                case 4:
                    sobj.deleteFirst();
                    break;
                case 5:
                    sobj.deleteLast();
                    break;
                case 6:
                    System.out.println("Enter position");
                    iPos = s.nextInt();
                    sobj.deleteAtpos(iPos);
                    break;
                case 7:
                    sobj.display();
                    System.out.println();
                    break;
                case 8:
                    iRet = sobj.count();
                    System.out.println("NUMBER OF ELEMENTS IN LINKED LIST :" + iRet);
                    break;
                case 9:
                    System.out.println("Existing program");
            }
        } while (ch != 9);

    }
}
